ddtrace/tracer/tracer.go

```diff
@@ -585,6 +653,11 @@ func (t *tracer) StartSpan(operationName string, options ...ddtrace.StartSpanOpt
 		log.Debug("Started Span: %v, Operation: %s, Resource: %s, Tags: %v, %v",
 			span, span.Name, span.Resource, span.Meta, span.Metrics)
 	}
+	if t.config.profilerHotspots || t.config.profilerEndpoints {
+		t.applyPPROFLabels(span.pprofCtxRestore, span)
+	} else {
+		span.pprofCtxRestore = nil
```

Comment:
I assume this is needed now because `SpanStart` unconditionally puts something in this field, right? But if I comment out this line, all the tests still pass. So I worry there is some untested behavior here.

Related, but is there a benefit to splitting the pprof label and execution trace task logic between these two span creation functions? GitHub won't let me comment on deleted lines, but I'm referring to the deleted block preceding this line. I think I prefer the way it was before where `startExecutionTracerTask` and `applyPPROFLabels` happen back to back. That way I find it's more clear that both things are trying to carry values through the same context. The context propagation logic is already tricky as it is (I struggle to wrap my head around it any time I need to work with it) and I worry it will become even more difficult to understand if it's spread out more in the code.

Reply:
We're trying to split the `Span` logic out of the `Tracer` as much as possible, so we can make `Spans` independent of the tracer. This way we can use Spans with things like the mock tracer or other potential implementations.

Unfortunately, there is high coupling there. Some things are tracer-specific, including this pprof stuff. It would be nice to make this independent of the tracer entirely, but as you said, the logic is complicated and I didn't want to change it any more than absolutely necessary to achieve separation.

Please let me know if you see a better way to handle this.

> I assume this is needed now because SpanStart unconditionally puts something in this field, right? But if I comment out this line, all the tests still pass. So I worry there is some untested behavior here.

Correct. You're probably right too about missing coverage. I'll look into that.

Reply:
> Correct. You're probably right too about missing coverage. I'll look into that.

I was able to exercise a difference with this test:

```go
func TestPprofLabels(t *testing.T) {
	if err := tracer.Start(
		tracer.WithProfilerCodeHotspots(false),
		tracer.WithProfilerEndpoints(false),
	); err != nil {
		t.Fatal(err)
	}
	defer tracer.Stop()
	pprof.Do(context.Background(), pprof.Labels("foo", "bar"), func(ctx context.Context) {
		wasteA(time.Second)
		var span *tracer.Span
		pprof.Do(ctx, pprof.Labels("foo", "baz"), func(ctx context.Context) {
			span, _ = tracer.StartSpanFromContext(ctx, "myoperation")
			wasteB(time.Second)
		})
		span.Finish()
		wasteC(time.Second)
	})
}
```

Here, `waste{A,B,C}` are just functions which waste the given amount of CPU time, so I can have some samples to distinguish which labels are active for the goroutine. I ran this with the CPU profiler enabled and looked at what labels are assigned to the `wasteC` samples. I would expect, with code hotspots and endpoint profiling disabled, that the `wasteC` samples have the label `"foo": "bar"`. Indeed, that's the case with this PR as-is and on main. If I comment out the `span.pprofCtxRestore = nil` case in this PR, the label for `wasteC` samples becomes `"foo": "baz"`. This helps convince me that this line is correct. Granted, this is a somewhat silly example. I don't know if a user would ever write code like this. But if we feel that it would be a good thing to have, I can turn it into a proper test (actually check the CPU profile output) as a followup.

Tangential, but I think as a user I might be surprised to find that with code hotspots and/or endpoint profiling _enabled_, the `wasteC` samples are tagged with `"foo": "baz"`. Naively reading the code, it wouldn't be obvious that the tracer functions are manipulating the labels at all. But, this PR is at least consistent with the current behavior, so... 🤷 The semantics of pprof labels don't really play well with these distributed tracing APIs, at least not nearly as well as execution tracer tasks do.

Our existing test suite passes otherwise, including traceproftest, so code hotspots & endpoint profiling seem to still work. The execution tracer tests here and in v2/contrib/database/sql pass as well (tried with Go 1.20). So it doesn't appear to be any more broken than it was before. With that in mind, I think I could tentatively approve this.

> Please let me know if you see a better way to handle this.

I don't have a better idea at the moment, unfortunately :/

If we want to feel more confident refactoring this code, I think we'd really benefit from nailing down the expected behavior with regard to context propagation, pprof labels, etc with good tests. I can tackle this, but I don't want to unduly block this work since I don't know when I'll have time to look into it. I do worry it will be even more difficult if we have to fix up two divergent versions of this code.